### DISCUSSION TITLE
Implement C's fstat in V for improved file status checking

### DISCUSSION THREAD

**Discussion by santerijps** [vlang/v#14670] [Ideas]
# Implement C's `fstat` function in V

Could C's `fstat` function be implemented in V? This seems to work better for getting the current size of a file being written to in parallel. Also, the struct returned by the function could be used for getting file creation, modification access times etc.

[Here's a link to the man page of fstat.](https://linux.die.net/man/2/fstat)

## System information

**OS name:** 64-bit Windows 10 Home
**OS version:** 10.0.19043 Build 19043
**V version:** 0.2.4 a3c0a9b

## Motivation

I have written a V program that makes an HTTP request to a remote file and downloads it while printing the download progress to the terminal. The body of the HTTP response is streamed with the `io` module function `cp` to a local file.  To keep track of the download progress, I'm running a goroutine that checks the size of the file every 500 milliseconds.

**If I use the current `os` module function `file_size`, the value returned is always 0**. How to get around this issue?

So I wrote some V code to implement a C-style solution (I believe Golang also uses `stat`):

## Example solution

To get around the issue, I wrote the code below. Someone wiser than me could explain why the below code works compared to the `os.file_size` function, as the `os.file_size`  [uses a similar C function `_wstat64` to get the file status.](https://github.com/vlang/v/blob/33a2d00445cb2a1e60ce4947337730152be6b4d4/vlib/os/os.c.v#L155) 

```v
#include <fcntl.h>
#include <sys/stat.h>

struct C.stat {
	st_dev		u64
	st_size		u64
	st_mode		u64
	st_mtime	u64
	st_ctime	u64
	st_atime	u64
}

fn C.fstat(int, &C.stat) int

pub fn fstat_fd(fd int) C.stat {
	s := C.stat{}
	C.fstat(fd, &s)
	return s
}

pub fn fstat_fn(path string) C.stat {
	fd := C.open(&char(path.str), C.O_RDONLY)
	s := C.stat{}
	C.fstat(fd, &s)
	C.close(fd)
	return s
}

```

_**Is this implementation reasonable? Is there some other way to work around the problem described in this issue?**_

**Comment by dumblob** [comment]
I would guess there is a subtle semantics difference to the POSIX standard allowing C stdlib to cache some of the results or calls.

But this is pure guesswork - I did not check neither the source of glibc nor POSIX standard nor any other lib.
